homie-controller/tests/integration.rs

```diff
@@ -0,0 +1,220 @@
+//! Integration test between homie-controller and homie-device crates. Starts a device and a
+//! controller connected to the same MQTT broker, and ensures that the controller can discover the
+//! device as expected.
+
+use futures::future::ready;
+use homie_controller::{Event, HomieController, State};
+use homie_device::{HomieDevice, Node, Property, SpawnError};
+use librumqttd::{Broker, Config};
+use rumqttc::{ConnectionError, MqttOptions, StateError};
+use std::io::ErrorKind;
+use std::sync::{Arc, Mutex};
+use std::thread;
+
+// A high port number which is hopefully not in use, to use for the MQTT broker.
+const PORT: u16 = 10883;
+
+#[tokio::test]
+async fn test_device() {
+    pretty_env_logger::init();
+
+    // Start MQTT broker.
+    spawn_mqtt_broker(PORT);
+
+    // Start controller.
+    let controller_options = MqttOptions::new("homie_controller", "localhost", PORT);
+    let (controller, mut event_loop) = HomieController::new(controller_options, "homie");
+    controller.start().await.unwrap();
```

Comment:
maybe make the function return Result and use ?

Reply:
It works in some cases, but in a bunch of places I'm unwrapping `Option`s, so it doesn't work there. There are also a few inside closures. I think it's more consistent just to use `unwrap()` everywhere, and I guess will also give more useful backtraces.